src/box/lua/config/instance_config.lua

```diff
@@ -484,6 +484,77 @@ return schema.new('instance_config', schema.record({
                 end
             end,
         })),
+        storage = enterprise_edition(schema.record({
+            prefix = schema.scalar({
+                type = 'string',
+                validate = function(data, w)
+                    if not data:startswith('/') then
+                        w.error(('config.tarantool.prefix should be ' ..
+                            'a path alike value, got %q'):format(data))
+                    end
+                end,
+            }),
+            endpoints = schema.array({
+                items = schema.record({
+                    uri = schema.scalar({
+                        type = 'string',
+                    }),
+                    login = schema.scalar({
+                        type = 'string',
+                    }),
+                    password = schema.scalar({
+                        type = 'string',
+                    }),
+                    params = schema.record({
+                        transport = schema.enum({
+                            'plain',
+                            'ssl',
+                        }),
+                        ssl_key_file = enterprise_edition(schema.scalar({
+                            type = 'string',
+                        })),
+                        ssl_cert_file = enterprise_edition(schema.scalar({
+                            type = 'string',
+                        })),
+                        ssl_ca_file = enterprise_edition(schema.scalar({
+                            type = 'string',
+                        })),
+                        ssl_ciphers = enterprise_edition(schema.scalar({
+                            type = 'string',
+                        })),
+                        ssl_password = enterprise_edition(schema.scalar({
+                            type = 'string',
+                        })),
+                        ssl_password_file = enterprise_edition(schema.scalar({
+                            type = 'string',
+                        })),
+                    }),
+                }),
+                validate = function(data, w)
+                    if #data == 0 then
+                        w.error('At least one endpoint must be' ..
+                            'specified in config.storage.endpoints')
+                    end
+                end,
+            }),
+            timeout = schema.scalar({
+                type = 'number',
+                default = 3,
+            }),
+            reconnect_after = schema.scalar({
+                type = 'number',
+                default = 3,
+            })
```

Comment:
It seems, if `config.storage` is missed in config and `:apply_default()` is called, `:validate()` doesn't pass on the resulting data. See a comment for `config.etcd`.

It seems, we're lucky that it is not called in this order on our testing scenarios.

We can either move defaults to the source (and leave them here as comments, just like in `config.etcd.http.request.timeout`. Or keep the defaults in the schema and change the `validate` annotation logic.

I would prefer the second. Let's just ensure that either `prefix` and `endpoints` are set both or unset both. It matches the source enabling criteria in the EE part: if `endpoints` is `nil` the source is disabled.

Reply:
Thank you! Checks added